examples/physics/raycast.rs

```diff
@@ -0,0 +1,146 @@
+        .run();
+}
+
+fn setup(
+    commands: &mut Commands,
+    mut meshes: ResMut<Assets<Mesh>>,
+    mut materials: ResMut<Assets<StandardMaterial>>,
+) {
+    commands
+        .spawn(PbrBundle {
+            mesh: meshes.add(Mesh::from(shape::Plane { size: 10.0 })),
+            material: materials.add(Color::rgb(0.1, 0.2, 0.1).into()),
+            ..Default::default()
+        })
+        .spawn(PbrBundle {
+            mesh: meshes.add(Mesh::from(shape::Icosphere {
+                subdivisions: 4,
+                radius: 0.5,
+            })),
+            material: materials.add(Color::rgb(0.1, 0.4, 0.8).into()),
+            transform: Transform::from_translation(Vec3::new(1.5, 1.5, 0.0)),
+            ..Default::default()
+        })
+        .spawn(LightBundle {
+            transform: Transform::from_translation(Vec3::new(4.0, 8.0, 4.0)),
+            ..Default::default()
+        })
+        .spawn(Camera3dBundle {
+            transform: Transform::from_translation(Vec3::new(-3.0, 5.0, 8.0))
+                .looking_at(Vec3::zero(), Vec3::unit_y()),
+            ..Default::default()
+        })
+        .insert_resource(MouseState::default());
+}
+
+#[derive(Default)]
+struct MouseState {
+    mouse_button_event_reader: EventReader<MouseButtonInput>,
+    cursor_moved_event_reader: EventReader<CursorMoved>,
+    cursor_position: Vec2,
+}
+
+fn raycast(
+    commands: &mut Commands,
+    mut meshes: ResMut<Assets<Mesh>>,
+    mut materials: ResMut<Assets<StandardMaterial>>,
+    mut mouse_state: ResMut<MouseState>,
+    mouse_button_input_events: Res<Events<MouseButtonInput>>,
+    cursor_moved_events: Res<Events<CursorMoved>>,
+    windows: Res<Windows>,
+    cameras: Query<(&Camera, &GlobalTransform)>,
+) {
+    for event in mouse_state
+        .cursor_moved_event_reader
+        .iter(&cursor_moved_events)
+    {
+        mouse_state.cursor_position = event.position;
+    }
+
+    for event in mouse_state
+        .mouse_button_event_reader
+        .iter(&mouse_button_input_events)
+    {
+        if event.button == MouseButton::Left && event.state == ElementState::Pressed {
+            for (camera, global_transform) in cameras.iter() {
+                let window = windows.get(camera.window).unwrap();
+
+                let ray = Ray::from_mouse_position(
+                    &mouse_state.cursor_position,
+                    window,
+                    camera,
+                    global_transform,
+                );
+
+                let plane_hit = Plane::new(Vec3::zero(), Vec3::unit_y()).intersect_ray(&ray);
+                let sphere_hit = Sphere {
+                    center: Vec3::new(1.5, 1.5, 0.0),
+                    radius: 0.5,
+                }
+                .intersect_ray(&ray);
+
+                let hit = if let Some(plane_hit) = plane_hit {
+                    if let Some(sphere_hit) = sphere_hit {
+                        if plane_hit.t() < sphere_hit.t() {
```

Comment:
We have 6 nested (for and if) statements here, that is very complex.

Reply:
Although I disagree with 6 nested blocks being inherently complex, I agree that this particular instance is unnecessarily verbose and doesn't get the point across. I'm looking for something more elegant but feel free to propose changes where you see fit.